README.md
# cassandrasimple
R package to connect to Cassandra and do simple operations

This package wraps the Datastax Cassandra Java Driver version 3.1 in R.

We use rJava to interface with Java.  Using the Jar files provided by Datastax, we initiate Cassandra sessions, and the run CQL queries.

For reading data, we inspect the resultset data types and construct data.frame with the appropriate column types and conversion of Cassandra NULL to R NA.

```
jCassSess <- get_cass_session('cassandracluster-internal.mycompany.com', 'myCluster', 'myKeyspace')

cass_update(jCassSess, "create table test_table (
    dt date,
    hr tinyint, 
    obj_id int, 
    val_1 float,
    val_2 float,
    primary key ((dt),hr,obj_id)  
)")

cass_update(jCassSess, "insert into test_table (dt,hr,obj_id,val_1,val_2) values ('2017-01-19',13,80,24.33,-34.33)")
cass_update(jCassSess, "BEGIN BATCH
      insert into test_table (dt,hr,obj_id,val_1,val_2) values ('2017-01-20',13,80,24.33,-34.33);
      insert into test_table (dt,hr,obj_id,val_1,val_2) values ('2017-01-20',24,80,14.44,55.55);
    APPLY BATCH")
    
stuff1 <- cass_query(jCassSess,"select * from test_table where dt = '2017-01-20' and hr = 13")    
# 
#           dt    hr obj_id val_1  val_2
# 1 2017-01-20    13     80 24.33 -34.33
# 2 2017-01-20    24     80 14.44  55.55

stuff2 <- cass_query(jCassSess,"select * from test_table where dt = '2017-01-20' and hr = 13")  
# 
#          dt    hr obj_id val_1  val_2
#1 2017-01-20    13     80 24.33 -34.33

close_cass_session(jCassSess)
```

To install this, do:

```
library(devtools)
install_github("thefooj/cassandrasimple")
```

We currently support these types:

* float => as.numeric
* double => as.numeric
* text => as.character
* varchar => as.character
* int => as.integer
* tinyint => as.integer
* date => as.Date.  Custom converter using CQL LocalDate to int days since 1970-01-01 and to asDate(x, origin='1970-01-01', tz='UTC')
* timestamp => as.POSIXct.  Custom converter using CQL Timestamp (milliseconds since epoch 1970-01-01) to long milliseconds -> seconds to asPOSIXct(x, tz='UTC', origin='1970-01-01')


== Writing Dataframes ==

As of version 1.0.1, we now support the batch-writing of data.frames to Cassandra.

```

# If you have a Cassandra table "test_table" with schema that looks like this:

cass_update(jCassSess, "create table if not exists test_insert_table (
              dt date,
              hr tinyint,
              obj_id int,
              int_1 int,
              int_2 int,
              float_1 float,
              float_2 float,
              double_1 double,
              str_val text,
              factor_str_val text,
              ts_val timestamp,
              dt_val date,
              primary key ((dt),hr,obj_id)
  )")
  
# And you have a data.frame:
myDf1 <- data.frame(dt=as.Date('2017-02-22'), hr=seq(1,24), obj_id=round(runif(24,min=1,max=1000),0),
                      int_1=as.integer(runif(24,min=1,max=24)),
                      int_2=as.numeric(seq(1,24)),
                      float_1=runif(24,min=-10,max=10),
                      float_2=runif(24,min=-1000,max=1000),
                      double_1=runif(24,min=-1000,max=1000),
                      str_val=paste("Hi! It's hour",seq(1,24)),
                      factor_str_val=as.factor(paste("fac_",round(runif(24,min=1,max=100),0),sep='')),
                      ts_val=as.POSIXct('2017-02-22 00:23:33',tz='EST'),
                      dt_val=as.Date('2017-03-01'),
                      stringsAsFactors = F
                      )

# ... and that data.frame has NAs                      
myDf1[3,]$float_1 <- NA
myDf1[3,]$str_val <- NA
myDf1[3,]$factor_str_val <- NA
myDf1[4,]$double_1 <- NA
myDf1[4,]$dt_val <- NA
myDf1[5,]$int_1 <- NA
myDf1[5,]$ts_val <- NA

# you can save it like this:
cass_save_df(jCassSess, myDf1, "test_insert_table", row_batches=2)

```

See `cass_save_df` for details.


== Running tests ==

First, you need to run `mkdist` from the command line.  This will create the Java classes needed.

Then in Rstudio you can Build & Reload and run tests




NEWS.md

## Change in Version 0.2.2 date 2018-06-19

Added support for boolean/logical type

## Changes in Version 0.2.1 date 2017-07-21

Added sleep time option to cass_save_df to sleep between batches


## Changes in Version 0.2.0 date 2017-03-30

Significant speed improvements.  Moved data fetching code out of rJava/JNI to pure Java.
In Java, we build up columnar data, and then quickly build up the data.frame from the list of columns.

Sample runtime improvements:


A query with 4446 rows:

```
[JavaBasedFaster... START @ 2017-03-30 14:15:47 ]
[JavaBasedFaster... STOP @ 2017-03-30 14:15:47 ] - got  4446  rows.

[rJavaSlower... START @ 2017-03-30 14:15:47 ]
[rJavaSlower... STOP @ 2017-03-30 14:15:58 ] - got  4446  rows.
```

A query with 177989 rows:

```
[JavaBasedFaster... START @ 2017-03-30 14:15:58 ]
[JavaBasedFaster... STOP @ 2017-03-30 14:16:03 ] - got  177989  rows.

[rJavaSlower... START @ 2017-03-30 14:16:03 ]
[rJavaSlower... STOP @ 2017-03-30 14:25:01 ] - got  177989  rows.
```




## Changes in Version 0.1.2 date 2017-02-14

Close down cluster connection in addition to session connection

## Changes in Version 0.1.1 date 2017-02-22

We added the capability to save a data.frame to Cassandra. via `cass_save_df`

## Initial Releave Version 0.1.0

This was our initial release


tests/testthat.R
library(testthat)
library(cassandrasimple)

test_check("cassandrasimple")


tests/testthat/test_connect.R
library(cassandrasimple)
library(dplyr)

# first, we must have Cassandra running locally, with keyspace 'cassandrasimple_test'
#  Start cassandra with
#     cd ~/dev/cassandra/apache-cassandra-3.10/bin
#     ./cassandra
#  Start CQLSH:
#     cd ~/dev/cassandra/apache-cassandra-3.10/bin
#     ./cqlsh
#  Set up keyspace:
#     create keyspace cassandrasimple_test with replication = {'class':'SimpleStrategy', 'replication_factor': 1};


test_that("connections work", {
  jCassSess = tryCatch({

    get_cass_session('localhost', 'Tios001', 'cassandrasimple_test')

  }, error = function(e) {
    cat("ERROR with Cassandra connection.  Be sure to have a local Cassandra running with cluster 'Tios001' and keyspace 'cassandrasimple_test'\n")
    cat("Suppose you have Cassandra downloaded in ~/dev/cassandra/apache-cassandra-3.10/bin\n")
    cat("Then run cassandra with:\n")
    cat("   cd ~/dev/cassandra/apache-cassandra-3.10/bin && ./cassandra\n")
    cat("Start CQLSH:\n")
    cat("   cd ~/dev/cassandra/apache-cassandra-3.10/bin && ./cqlsh\n")
    cat("Create the keyspace in CQLSH:\n")
    cat("   create keyspace cassandrasimple_test with replication = {'class':'SimpleStrategy', 'replication_factor': 1};\n")
  })

  close_cass_session(jCassSess)
})


tests/testthat/test_queries.R
library(cassandrasimple)
library(dplyr)

test_that("updates and queries work, converting NULL to NA and handling mappings for our supported types", {


  jCassSess <- get_cass_session('localhost', 'Test Cluster', 'cassandrasimple_test')

  cass_update(jCassSess, "drop table if exists test_table")

  cass_update(jCassSess, "create table if not exists test_table (
              dt date,
              hr tinyint,
              obj_id int,
              float_1 float,
              float_2 float,
              double_1 double,
              str_val text,
              primary key ((dt),hr,obj_id)
  )")

  res <- cass_query(jCassSess, "SELECT * FROM test_table")
  expect_is(res, "data.frame")
  expect_equal(nrow(res),0)

  cass_update(jCassSess, "insert into test_table (dt,hr,obj_id,float_1,float_2,double_1,str_val) values ('2017-01-19',13,80,24.33,-34.33,0.33314,'HELLO THERE HOW ARE YOU')")
  cass_update(jCassSess, "BEGIN BATCH
              insert into test_table (dt,hr,obj_id,float_1,float_2,double_1,str_val) values ('2017-01-20',13,80,NULL,-34.33,NULL,NULL);
              insert into test_table (dt,hr,obj_id,float_1,float_2,double_1,str_val) values ('2017-01-20',24,80,14.44,55.55,8.2492994,'ANOTHER');
              APPLY BATCH")

  #           dt    hr obj_id double_1 float_1 float_2                 str_val
  #       (date) (int)  (int)    (dbl)   (dbl)   (dbl)                   (chr)
  # 1 2017-01-20    13     80       NA      NA  -34.33                      NA
  # 2 2017-01-20    24     80 8.249299   14.44   55.55                 ANOTHER
  # 3 2017-01-19    13     80 0.333140   24.33  -34.33 HELLO THERE HOW ARE YOU

  res <- cass_query(jCassSess, "SELECT * FROM test_table")
  expect_is(res, "data.frame")
  expect_equal(nrow(res),3)

  expect_equal(res %>% filter(dt == '2017-01-20' & hr == 24 & obj_id ==80) %>% select(str_val) %>% as.character(), 'ANOTHER')
  expect_true(abs(res %>% filter(dt == '2017-01-20' & hr == 24 & obj_id ==80) %>% select(float_1) %>% as.numeric() - 14.44) < 0.0001)
  expect_true(abs(res %>% filter(dt == '2017-01-20' & hr == 13 & obj_id ==80) %>% select(float_2) %>% as.numeric() - -34.33) < 0.0001)

  # int handling
  expect_equal(as.character(format(res %>% filter(dt == '2017-01-20' & hr == 24 & obj_id ==80) %>% select(hr) )), '24')

  # date handling
  expect_equal(as.character(format(res %>% filter(dt == '2017-01-20' & hr == 24 & obj_id ==80) %>% select(dt) )), '2017-01-20')

  # verify NULL cases
  expect_true(is.na((res %>% filter(dt == '2017-01-20' & hr == 13 & obj_id ==80) %>% select(str_val))$str_val))
  expect_true(is.na((res %>% filter(dt == '2017-01-20' & hr == 13 & obj_id ==80) %>% select(float_1))$float_1))
  expect_true(is.na((res %>% filter(dt == '2017-01-20' & hr == 13 & obj_id ==80) %>% select(double_1))$double_1))

  close_cass_session(jCassSess)
})

test_that("timestamp types are retrieved as UTC with appropriate conversions", {
  jCassSess <- get_cass_session('localhost', 'Test Cluster', 'cassandrasimple_test')

  cass_update(jCassSess, "drop table if exists test_timestamp_table")

  cass_update(jCassSess, "create table if not exists test_timestamp_table (
              dt date,
              hr tinyint,
              obj_id int,
              update_version_time timestamp,
              float_1 float,
              float_2 float,
              double_1 double,
              str_val text,
              primary key ((dt),hr,obj_id, update_version_time)
  )")

  # inserts based on discernable strings for timezone offsets
  cass_update(jCassSess, "BEGIN BATCH
              insert into test_timestamp_table (dt,hr,obj_id,update_version_time,float_1,float_2,double_1,str_val) values ('2017-01-20',13,80,'2017-01-19 07:05:02+0000',NULL,-34.33,NULL,NULL);
              insert into test_timestamp_table (dt,hr,obj_id,update_version_time,float_1,float_2,double_1,str_val) values ('2017-01-20',24,80,'2017-01-19 02:04:02-0500',14.44,55.55,8.2492994,'ANOTHER');
              APPLY BATCH")

  res <- cass_query(jCassSess, "SELECT * FROM test_timestamp_table")
  expect_is(res, "data.frame")
  expect_equal(nrow(res),2)
  expect_is(res$update_version_time, 'POSIXct')
  # converted over to UTC
  expect_equal(strftime(res$update_version_time, '%Y-%m-%d %H:%M:%S %Z', tz="UTC"),
               c('2017-01-19 07:05:02 UTC','2017-01-19 07:04:02 UTC'))


  close_cass_session(jCassSess)
})

R/save_df.R

.coerce_df_to_chars <- function(df) {
  dfCopy <- df
  for (x in colnames(df)) {
    #cat("Coercing to char: ",x,"\n")

    colClass = class(df[,x])
    if (length(colClass) > 1) {
      colClass = colClass[1]
    }

    # special case of all na
    if (all(is.na(dfCopy[,x]))) {
      dfCopy[,x] <- 'NULL'
    } else if (colClass=='numeric') {
      dfCopy[,x] <- as.character(round(df[,x],5))  # use 5 for now so we don't over-push details.  By rounding, any integer values will show as integers (3.000 -> 3)
    } else if(colClass=='character' || colClass=='factor') {
      dfCopy[,x] <- paste("'",gsub("'","''",as.character(df[,x]),fixed=T),"'",sep='')
    } else if (colClass=='Date') {
      dfCopy[,x] <- paste("'",as.character(df[,x]),"'",sep='')
    } else if (colClass=='POSIXct') {
      dfCopy[,x] <- format(df[,x], "'%Y-%m-%dT%H:%M:%S+0000'",tz='UTC')
    } else if (colClass=='integer') {
      dfCopy[,x] <- as.character(df[,x])
    } else if (colClass=='logical') {  # boolean T/F/NA
      dfCopy[,x] <- sapply(df[,x], function(x) { ifelse(is.na(x), NA,  ifelse(x,'true','false') ) })
    } else {
      stop("unknown df class for ", x, " (",colClass,")")
    }
    dfCopy[which(is.na(df[,x])),x] <- 'NULL'
  }
  return(dfCopy)
}

# returns vector of CQL of strings that are CQL INSERT statements
.df_to_cql <- function(df, tbl) {
  df_chars <- .coerce_df_to_chars(df=df)

  cols_csv <- paste(colnames(df),collapse=',',sep='')
  insert_head <- paste0("INSERT INTO ",tbl," (",cols_csv, ") VALUES")

  paste(insert_head, " (",apply(df_chars,1,function(row) {
    paste(row,collapse=',',sep='')
  }), ");\n",sep='')
}


#' Save a dataframe to Cassandra for the given table name
#' We currently handle:
#'   numeric -> float (autoconvert to int).  Floats are rounded to 5 digits currently
#'   date -> 'YYYY-mm-dd' strings
#'   POSIXct -> 'YYYY-mm-ddTHH:MM:SS+0000' in UTC.
#'   logical -> 1, 0, or NA.  If a column is all-NA, then the data.frame thinks it's 'logical'.  We save it as NA.
#'   character -> strings enclosed in single quotes, and escaped the CQL way.  e.g. 'It''s my party'
#'   all-NA  -> NA
#' NAs are represented as NULL in Cassandra, so we handle that correctly
#' The user of this function must specify a reasonable row_batches to prevent write failures.
#' If the batch is too large, you will bump against the batch_size_fail_threshold_in_kb limit in Cassandra (default 50kb)
#' If the batch is too small, the writes will take too long
#'
#' @param jCassSess The Cassandra session, from get_cass_session
#' @param df Data.frame.  Columns must match columns in the Cassandra table.  Otherwise the save will fail
#' @param table_name Name of the Cassandra table.  Save will fail if table does not exist
#' @param row_batches Number of inserts to run in each batch.  Batches are wrapped in BEGIN BATCH ... APPLY BATCH.  This implies logging.
#' @param sleep_between_batches Number of seconds to sleep between batches.  Defaults to 0.
cass_save_df <- function(jCassSess, df, table_name, row_batches = 1000, sleep_between_batches = 0, show_debug=F) {

  # do not use this function when you are working with a df that contains over 100,000 rows

  if (missing(table_name) | is.null(table_name) | identical(table_name, NA)) {
    stop('Invalid table name!')
  }

  if (!grepl('^[0-9a-z\\_]+$', table_name, perl = TRUE)) {
    stop('Illegal characters in the table name!')
  }

  if (sum(grepl('^[0-9a-z\\_]+$', names(df), perl = TRUE)) != ncol(df)) {
    stop('Illegal characters in the df column names!')
  }

  if(is.null(row_batches) | identical(row_batches, NA) | !identical(class(row_batches), 'numeric')) {
    stop('Invalid row batches!')
  }

  if (nrow(df) == 0) {
    cat("cass_save_df called with empty data frame.  Doing nothing")
    return(T)
  }


  batch_prefix <- 'BEGIN BATCH '
  batch_suffix <- ' APPLY BATCH'

  #each time, we only upload up to 1000 (row_batches) rows of data with one CQL insert command.
  #separate data frame into small group with no more than 1000 rows of data
  num_uploads <- ceiling(nrow(df) / row_batches)
  options(digits.secs=6)

  tot_prep_time <- 0
  tot_upload_time <- 0
  tot_sleep_time <- 0

  for (i in 1:num_uploads) {
    prep_start_time <- Sys.time()

    if (i*row_batches <= nrow(df)) {
      df_subset <- dplyr::slice(df, ((i - 1)*row_batches + 1):(i*row_batches))
    } else {
      df_subset <- dplyr::slice(df, ((i - 1)*row_batches + 1):nrow(df))
    }

    insert_command_line <- .df_to_cql(df_subset, table_name)

    combined_result <- paste0(batch_prefix, paste0(insert_command_line, collapse = ""), batch_suffix)

    prep_stop_time <- Sys.time()

    cass_update(jCassSess, cqlQuery = combined_result)

    upload_stop_time <- Sys.time()

    if (show_debug) {
      cat("[SaveDfToCassandra:",table_name," batch=",i," @", Sys.time(), " - prep: ", (prep_stop_time - prep_start_time), "   upload: ", (upload_stop_time - prep_stop_time),"\n")
    }

    tot_prep_time <- tot_prep_time + (prep_stop_time - prep_start_time)
    tot_upload_time <- tot_upload_time + (upload_stop_time - prep_stop_time)

    if (sleep_between_batches > 0) {
      if (show_debug) {
        cat("[SaveDfToCassandra:",table_name," sleep ", sleep_between_batches,"\n")
      }
      Sys.sleep(sleep_between_batches)
    }

    tot_sleep_time <- tot_sleep_time + sleep_between_batches
  }

  options(digits.secs=NULL)

  if (show_debug) {
    cat("Total prep: ",tot_prep_time,"\n")
    cat("Total upload: ", tot_upload_time,"\n")
    cat("Total sleep: ", tot_sleep_time,"\n")

    cat('We have uploaded', i, 'batches to Cassandra.\n')
  }

  return(T)
}


R/session.R
#' Get a Cassandra session.  Returns a com.datastax.driver.core.Session session reference.
#' When you run this, you will get warnings about failed logger loading.
#'
#' @example jCassSess <- get_cass_session('my.ip', 'myCluster', 'myKeyspace')
#'
#' @param connectUrl IP address or hostname.  If the host name points to a DNS record with multiple a-records, all addresses will be used by the Datastax driver.  See: http://docs.datastax.com/en/drivers/java/3.0/com/datastax/driver/core/Cluster.Builder.html#addContactPoint-java.lang.String-
#' @param clusterName The name of the Cassandra cluster
#' @param clusterKeyspace The name of the Cassandra keyspace in the cluster
#' @export
get_cass_session <- function(connectUrl, clusterName, clusterKeyspace,debug=F) {

  if (debug==T) { cat("get_cass_session:: ... setting classpath, java init, and loading jars\n")}
  jarBase <- system.file("java", package='cassandrasimple')
  rJava::.jinit()
  rJava::.jaddClassPath(Sys.glob(paste0(jarBase,"/*.jar")))

  jCassSess = rJava::.jcall('com.tioscapital.cassandrasimple.CassandraConnector', "Lcom/datastax/driver/core/Session;", 'getConnection', connectUrl, clusterName, clusterKeyspace, debug)

  # if (debug==T) { cat("get_cass_session:: ... initializing cluster\n")}
  # jClusterBuilder = rJava::.jcall('com.datastax.driver.core.Cluster', "Lcom/datastax/driver/core/Cluster$Builder;", 'builder')
  # if (debug==T) { cat("get_cass_session:: ... setting cluster name\n")}
  # jClusterBuilder = rJava::.jcall(jClusterBuilder,  "Lcom/datastax/driver/core/Cluster$Builder;", 'withClusterName', clusterName)
  # if (debug==T) { cat("get_cass_session:: ... adding contact point\n")}
  # jClusterBuilder = rJava::.jcall(jClusterBuilder,  "Lcom/datastax/driver/core/Cluster$Builder;", 'addContactPoint', connectUrl)
  # if (debug==T) { cat("get_cass_session:: ... building cluster\n")}
  # jCluster = rJava::.jcall(jClusterBuilder,  "Lcom/datastax/driver/core/Cluster;", 'build')
  # if (debug==T) { cat("get_cass_session:: ... getting session\n")}
  # jCassSess = rJava::.jcall(jCluster, 'Lcom/datastax/driver/core/Session;', 'connect', clusterKeyspace)
  # if (debug==T) { cat("get_cass_session:: ... DONE\n")}
  return(jCassSess)
}

#' Close down a Cassandra session
#'
#' @param jCassSess Java com.datastax.driver.core.Session session, returned from get_cass_session
#' @export
close_cass_session <- function(jCassSess,debug=F) {
  if (debug==T) { cat("close_cass_session:: ... closing Cassandra session\n") }
  jCluster = rJava::.jcall(jCassSess, "Lcom/datastax/driver/core/Cluster;", 'getCluster')
  rJava::.jcall(jCassSess, "V", 'close')
  if (debug==T) { cat("close_cass_session:: ... closing Cassandra cluster\n") }
  rJava::.jcall(jCluster, "V", 'close')
  if (debug==T) { cat("close_cass_session:: ... closing complete for Cassandra close_cass_sess\n") }
  return(T)
}

java-src/CassandraReader.java
package com.tioscapital.cassandrasimple;

import com.datastax.driver.core.Session;
import com.datastax.driver.core.ResultSet;
import com.datastax.driver.core.Row;
import com.datastax.driver.core.ColumnDefinitions;
import com.datastax.driver.core.DataType;
import java.util.ArrayList;
import java.util.List;
import java.util.Hashtable;
import java.util.Date;
import com.tioscapital.cassandrasimple.CassandraColumn;


public class CassandraReader {





  List<CassandraColumn> columns;
  Hashtable<String, Integer> colNameToIndexLookup;


  public CassandraReader(Session cassSess, String cqlQuery, int estimatedRows, boolean debug) {
    ResultSet rs =  cassSess.execute(cqlQuery);


    if (debug) { System.out.println("DONE with query... now getting column mappings... "); }
    this.initializeColumnMappings(rs, estimatedRows, debug);
    if (debug) { System.out.println("Column Mapping: " + colNameToIndexLookup.toString()); }

    if (debug) {
      for (int i = 0; i < columns.size(); i++) {
        ((CassandraColumn)(columns.get(i))).printDebug();
      }
    }

    if (debug) { System.out.println("START fetching data in Java"); }
    this.loadData(rs);
    if (debug) { System.out.println("DONE fetching data in Java"); }
  }

  public int getNumColumns() {
    return(columns.size());
  }

  /* i is the 0-based index */
  public CassandraColumn getColumn(int i) {
    return((CassandraColumn)(columns.get(i)));
  }

  private void loadData(ResultSet rs) {
    int i;
    int rowNum = 0;
    while (!rs.isExhausted()) {
      rowNum++;
      Row row = rs.one();
      for (i = 0; i < columns.size(); i++) {
        ((CassandraColumn)(columns.get(i))).addFromRow(row);
      }
    }
  }

  private void initializeColumnMappings(ResultSet rs, int estimatedRows, boolean debug) {
    ColumnDefinitions cdefs = rs.getColumnDefinitions();
    int numCols = cdefs.size();
    String cName;

    this.columns = new ArrayList<CassandraColumn>(numCols);
    this.colNameToIndexLookup = new Hashtable<String, Integer>();

    for (int i=0; i < numCols; i++) {
      cName = cdefs.getName(i);
      if (debug) { System.out.println("Got col " + cName); }

      // String colName, String colCassTypeStr, int cassColIdx, int estimatedSize
      CassandraColumn col = new CassandraColumn(cName, cdefs.getType(i).asFunctionParameterString(), i, estimatedRows, debug);
      if (debug) { System.out.println("Done with col setup: " + cName); }
      this.columns.add(col);
      this.colNameToIndexLookup.put(cName, i);
    }
  }


}

java-src/CassandraConnector.java
package com.tioscapital.cassandrasimple;

import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.Session;
import com.datastax.driver.core.PoolingOptions;
import com.datastax.driver.core.HostDistance;

public class CassandraConnector {

  public static com.datastax.driver.core.Session getConnection(String connectionUrl, String clusterName, String clusterKeyspace, boolean debug) {

    if (debug) System.out.println("CassandraConnector: getConnection -- Got session...");

    /* one connection only for simplicity */
    //PoolingOptions poolingOpts = new PoolingOptions();
    //poolingOpts.setConnectionsPerHost(HostDistance.LOCAL, 1, 256);
    //poolingOpts.setConnectionsPerHost(HostDistance.REMOTE, 1, 256);
    // .withPollingOptions(pollingOpts)


    Cluster cluster = Cluster.builder()
      .withClusterName(clusterName)
      .addContactPoint(connectionUrl)
      .build();

    //System.out.println("max_connections_per_host(LOCAL):" + cluster.getConfiguration().getPoolingOptions().getMaxConnectionsPerHost(HostDistance.LOCAL));
    //System.out.println("max_connections_per_host(REMOTE):" + cluster.getConfiguration().getPoolingOptions().getMaxConnectionsPerHost(HostDistance.REMOTE));
    //System.out.println("max_queue_size" + cluster.getConfiguration().getPoolingOptions().getMaxQueueSize());

    if (debug) System.out.println("CassandraConnector: getConnection -- cluster setup complete.  Getting session...");

    Session sess = cluster.connect(clusterKeyspace);

    if (debug)  System.out.println("CassandraConnector: getConnection -- Got session...");

    return(sess);
  }
}

R/query.R

#' for a given Java com.datastax.driver.core.ResultSet object, get the column definitions as a List.
#' Raises errors if we do not yet support the column type.
#' Currently we support: float, double, text, varchar, int, tinyint, date (assumes UTC context), timestamp (stored as POSIXct in timezone UTC)
#'
#' @param jcRes instance of com.datastax.driver.core.ResultSet
.cass_col_map <- function(jcRes) {

  jcColDefs <- rJava::.jcall(jcRes, 'Lcom/datastax/driver/core/ColumnDefinitions;', 'getColumnDefinitions')

  numCols <- rJava::.jcall(jcColDefs, 'I', 'size')

  mappings = list()
  for (i in 1:numCols) {
    cname <- rJava::.jcall(jcColDefs, "S", "getName", as.integer(i-1))
    ctypeObj <- rJava::.jcall(jcColDefs, 'Lcom/datastax/driver/core/DataType;','getType', as.integer(i-1))
    ctype <- rJava::.jcall(ctypeObj, 'S', 'asFunctionParameterString')

    map = list()
    map$index = as.integer(i-1)
    map$javatype = ctype
    map$colname = cname

    if (ctype=='float') {
      map$java_get_func = 'getFloat'
      map$r_cast_func   = 'as.numeric'
      map$jni_type      = 'F'
    } else if (ctype=='double') {
      map$java_get_func = 'getDouble'
      map$r_cast_func   = 'as.numeric'
      map$jni_type      = 'D'
    } else if (ctype=='text'||ctype=='varchar') {
      map$java_get_func = 'getString'
      map$r_cast_func   = 'as.character'
      map$jni_type      = 'S'
    } else if (ctype=='int') {
      map$java_get_func = 'getInt'
      map$r_cast_func   = 'as.integer'
      map$jni_type      = 'I'
    } else if (ctype=='tinyint') {
      map$java_get_func = 'getByte'
      map$r_cast_func   = 'as.integer'
      map$jni_type      = 'B'
    } else if (ctype=='date') {
      map$java_get_func = 'getDate'
      map$r_cast_func   = 'as_date_from_cql_date'  # our custom function
      map$jni_type      = 'Lcom/datastax/driver/core/LocalDate;'
    } else if (ctype=='timestamp') {
      map$java_get_func = 'getTimestamp'
      map$r_cast_func   = 'as_posixct_from_cql_get_timestamp'
      map$jni_type      = 'Ljava/util/Date;'
    } else {
      stop(paste0("Unsupported type of column ",cname," type ",ctype," in query "))
    }
    mappings[[cname]] <- map
  }
  return(mappings)
}



#' Perform a CQL update.
#'
#' @param jCassSess The Cassandra session, from get_cass_session
#' @param cqlQuery A string query.  If performing multiple at once, use BEG
#'
#' @examples
#'   cass_update(jCassSess, "insert into test_lmp_combined (dt,hr,official_pnode_id,rt_energy,da_energy) values ('2017-01-19',13,80,24.33,-34.33)")
#'   cass_update(jCassSess, "BEGIN BATCH
#'       insert into test_lmp_combined (dt,hr,official_pnode_id,rt_energy,da_energy) values ('2017-01-20',13,80,24.33,-34.33);
#'       insert into test_lmp_combined (dt,hr,official_pnode_id,rt_energy,da_energy) values ('2017-01-20',24,80,14.44,55.55);
#'       APPLY BATCH")
#' @export
cass_update <- function(jCassSess, cqlQuery) {
  rJava::.jcall(jCassSess, 'Lcom/datastax/driver/core/ResultSet;', 'execute', cqlQuery)
  return(T)
}

cass_query <- function(jCassSess, cqlQuery, estimatedNumRows=1000, debug=F) {
  reader <- rJava::.jnew("com/tioscapital/cassandrasimple/CassandraReader",jCassSess, cqlQuery, as.integer(estimatedNumRows), debug);
  numCols <- rJava::.jcall(reader, 'I', 'getNumColumns')

  dfList <- list()

  if (debug) { cat("... in new cass_query I have ",numCols," columns\n") }
  for (i in seq.int(numCols)) {
    col <- rJava::.jcall(reader, 'Lcom/tioscapital/cassandrasimple/CassandraColumn;', 'getColumn', as.integer(i-1))
    colName <- rJava::.jcall(col, 'S', 'getColName')
    colGetter <- rJava::.jcall(col, 'S', 'getGetMethodName')
    colGetterJniType <- rJava::.jcall(col, 'S', 'getGetMethodJniType')
    colGetterRConversion <- rJava::.jcall(col, 'S', 'getRConversionName')

    colData <- rJava::.jcall(col, colGetterJniType, colGetter)
    nullIndexes <- rJava::.jcall(col, "[I", "getNullIndexes")

    if (nchar(colGetterRConversion) > 0) {
      colData <- get(colGetterRConversion)(colData)
    }

    if (length(nullIndexes) > 0) {
      nullIndexes <- nullIndexes + 1  # indexes in java are 0-based
      if (debug) { cat("for ", colName, " the null indexes are ", paste0(nullIndexes,collapse=","), "\n") }
      colData[nullIndexes] <- NA
    }
    dfList[[colName]] <- colData
  }
  df <- as.data.frame(dfList, stringsAsFactors=F)
  return(df)
}

#' Execute a CQL query and capture the results as a data.frame.  Columns will match up with what is in the Cassandra table definition and your query.
#' Follow CQL rules for queries.  If you try to query on a column without a secondary index, you will get errors from the Datastax driver.
#' NULL values in the Cassandra tables are returned as NA in the data frame.  We convert Cassandra data types to reasonable R equivalents.
#' We currently support the following mappings:
#'   float, double, text, varchar, int, tinyint, date
#' See .cass_col_map for mapping details.
#'
#' @example stuff <- cass_query(jCassSess,"select * from test_lmp_combined where dt = '2017-01-20' and hr = 13")
#' @param jCassSess The Cassandra session, from get_cass_session
#' @param cqlQuery A string query.
#' @export
cass_query_slower <- function(jCassSess, cqlQuery) {
  jcRes <- rJava::.jcall(jCassSess, 'Lcom/datastax/driver/core/ResultSet;', 'execute', cqlQuery)

  results = list()

  colMaps <- .cass_col_map(jcRes)

  ctr <- 0
  while(!rJava::.jcall(jcRes, 'Z', 'isExhausted')) {
    jcRow <- rJava::.jcall(jcRes, 'Lcom/datastax/driver/core/Row;', 'one')
    ctr <- ctr+1
    row = list()
    for (m in names(colMaps)) {
      if (rJava::.jcall(jcRow, 'Z', "isNull", m)) {
        val <- NA
      } else {
        val <- rJava::.jcall(jcRow, colMaps[[m]]$jni_type, colMaps[[m]]$java_get_func, m)
        val <- get(colMaps[[m]]$r_cast_func)(val)
      }
      row[[m]] <- val
    }
    results[[ctr]] <- row
  }
  return(dplyr::bind_rows(results))
}